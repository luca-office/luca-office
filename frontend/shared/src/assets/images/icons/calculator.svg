<?xml version="1.0" encoding="UTF-8"?>
<svg viewBox="0 0 16 16" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <g id="Icons---lined" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd" stroke-linecap="round" stroke-linejoin="round">
        <g id="6_Customizable-Icons/calculator" stroke="currentColor">
            <path d="M1.5,14.5 L14.5,14.5 L14.5,1.5 L1.5,1.5 L1.5,14.5 Z M3.75,3.75 L6.25,6.25 M6.25,3.75 L3.75,6.25 M6.25,12.25 L6.25,12.25 M6.25,9.75 L3.75,12.25 M11,3.5 L11,6.5 M12.5,5 L9.5,5 M3.75,9.75 L3.75,9.75 M12.5,11 L9.5,11" id="Stroke-2"></path>
        </g>
    </g>
</svg>
