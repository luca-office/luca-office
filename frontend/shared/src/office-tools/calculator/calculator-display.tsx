import {css} from "@emotion/react"
import * as React from "react"
import {useClipboard} from "use-clipboard-copy"
import {Text, Tooltip} from "../../components"
import {errorColor, Flex, primaryColor, spacingMedium, spacingSmall, textEllipsis, TextSize} from "../../styles"
import {useLucaTranslation} from "../../translations"
import {Option} from "../../utils"

export interface CalculatorDisplayProps {
  readonly operation: Option<string>
  readonly result: Option<string>
  readonly hasInvalidInput: boolean
}

export const CalculatorDisplay: React.FC<CalculatorDisplayProps> = ({operation, result, hasInvalidInput}) => {
  const {t} = useLucaTranslation()

  const {copy} = useClipboard()
  const onResultClick = () => !hasInvalidInput && result.forEach(copy)

  return (
    <div css={[Flex.column, styles.container]}>
      <Text customStyles={styles.operation}>{operation && operation.orNull()}</Text>
      <Tooltip title={t("calculator__copy_to_clipboard")} inactive={hasInvalidInput} withPortal={true}>
        <div css={styles.resultRow(hasInvalidInput)} onClick={onResultClick}>
          <Text customStyles={styles.resultIcon} className={"result"}>
            =
          </Text>

          <Text customStyles={styles.result(hasInvalidInput)} className={"result"}>
            {hasInvalidInput ? t("calculator__invalid_input") : result && result.orNull()}
          </Text>
        </div>
      </Tooltip>
    </div>
  )
}

const styles = {
  container: css({
    marginBottom: spacingMedium
  }),
  operation: css(textEllipsis, {
    marginBottom: spacingSmall,
    fontSize: 24,
    letterSpacing: 0.23,
    height: 31,
    textAlign: "right"
  }),
  resultIcon: css({
    flex: 0,
    marginRight: spacingMedium,
    fontSize: TextSize.Huge,
    letterSpacing: 0.34
  }),
  resultRow: (isInvalid: boolean) =>
    css(Flex.row, {
      cursor: isInvalid ? "default" : "pointer",
      ...(!isInvalid && {
        "&:hover > .result": {
          color: primaryColor
        }
      })
    }),
  result: (isInvalid: boolean) =>
    css(textEllipsis, {
      flex: 1,
      fontSize: isInvalid ? TextSize.Large : TextSize.Huge,
      letterSpacing: 0.34,
      textAlign: "right",
      color: isInvalid ? errorColor : undefined
    })
}
