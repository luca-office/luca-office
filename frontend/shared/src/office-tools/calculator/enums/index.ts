export * from "./calculator-button-type"
export * from "./calculator-key"
