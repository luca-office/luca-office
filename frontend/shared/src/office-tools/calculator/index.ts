export * from "./calculator"
export * from "./calculator-container"
export * from "./enums"
export * from "./utils/calculator-utils"
