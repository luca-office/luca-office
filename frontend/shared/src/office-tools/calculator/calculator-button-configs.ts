import {css} from "@emotion/react"
import {CustomStyle} from "../../styles"
import {CalculatorButtonType} from "./enums/calculator-button-type"
import {CalculatorKey} from "./enums/calculator-key"

export interface CalculatorButtonConfig extends CustomStyle {
  readonly calculatorKey: CalculatorKey
  readonly type?: CalculatorButtonType
}

export const calculatorButtonConfigs: CalculatorButtonConfig[] = [
  {calculatorKey: CalculatorKey.OpenBracket, customStyles: css({gridColumn: "1 / 3"})},
  {calculatorKey: CalculatorKey.CloseBracket, customStyles: css({gridColumn: "3 / 5"})},
  {calculatorKey: CalculatorKey.AC},

  {calculatorKey: CalculatorKey.Seven, type: CalculatorButtonType.Secondary},
  {calculatorKey: CalculatorKey.Eight, type: CalculatorButtonType.Secondary},
  {calculatorKey: CalculatorKey.Nine, type: CalculatorButtonType.Secondary},
  {calculatorKey: CalculatorKey.Divide},
  {calculatorKey: CalculatorKey.Reciprocal},

  {calculatorKey: CalculatorKey.Four, type: CalculatorButtonType.Secondary},
  {calculatorKey: CalculatorKey.Five, type: CalculatorButtonType.Secondary},
  {calculatorKey: CalculatorKey.Six, type: CalculatorButtonType.Secondary},
  {calculatorKey: CalculatorKey.Multiply},
  {calculatorKey: CalculatorKey.Exponentiation},

  {calculatorKey: CalculatorKey.One, type: CalculatorButtonType.Secondary},
  {calculatorKey: CalculatorKey.Two, type: CalculatorButtonType.Secondary},
  {calculatorKey: CalculatorKey.Three, type: CalculatorButtonType.Secondary},
  {calculatorKey: CalculatorKey.Subtraction},
  {calculatorKey: CalculatorKey.Result, customStyles: css({gridColumn: "5 / 6", gridRow: "4 / 6"})},

  {calculatorKey: CalculatorKey.Zero, type: CalculatorButtonType.Secondary, customStyles: css({gridColumn: "1 / 3"})},
  {calculatorKey: CalculatorKey.Separator, type: CalculatorButtonType.Secondary},
  {calculatorKey: CalculatorKey.Addition}
]
