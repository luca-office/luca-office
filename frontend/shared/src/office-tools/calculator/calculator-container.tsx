import * as React from "react"
import {OfficeTool, OfficeWindowType} from "../../enums"
import {CustomStyle} from "../../styles"
import {last} from "../../utils"
import {Calculator} from "./calculator"
import {CalculatorKey} from "./enums/calculator-key"

export interface CalculatorState {
  readonly openWindows: Array<OfficeWindowType>
  readonly minimizedWindows: Array<OfficeWindowType>
}

export interface CalculatorSurveyEvents {
  readonly sendCalculatorKeyPressedEvent: (key: CalculatorKey) => void
}

export interface CalculatorContainerProps extends CustomStyle {
  readonly onClose: () => void
  readonly onMinimize?: () => void
  readonly onOutsideClick?: () => void
  readonly useState: () => CalculatorState
  readonly useSurveyEvents: () => CalculatorSurveyEvents
}

export const CalculatorContainer: React.FC<CalculatorContainerProps> = ({
  customStyles,
  onClose,
  onMinimize,
  onOutsideClick,
  useState,
  useSurveyEvents
}) => {
  const {minimizedWindows, openWindows} = useState()
  const {sendCalculatorKeyPressedEvent} = useSurveyEvents()
  const [officeWindowHasFocus, setOfficeWindowHasFocus] = React.useState(true)

  const onKeyPress = (key: CalculatorKey) => sendCalculatorKeyPressedEvent(key)

  const handleOutsideClick = () => {
    if (openWindows.includes(OfficeTool.Calculator) && !minimizedWindows.includes(OfficeTool.Calculator)) {
      setOfficeWindowHasFocus(false)
      onOutsideClick?.()
    }
  }

  const handleOnMouseDown = () => setOfficeWindowHasFocus(true)

  const isFocussed =
    officeWindowHasFocus &&
    last(openWindows.filter(window => !minimizedWindows.includes(window)))
      .map(window => window === OfficeTool.Calculator)
      .getOrElse(false)

  return (
    <Calculator
      customStyles={customStyles}
      onClose={onClose}
      onKeyPress={onKeyPress}
      onMinimize={onMinimize}
      onOutsideClick={handleOutsideClick}
      onMouseDown={handleOnMouseDown}
      isFocussed={isFocussed}
    />
  )
}
