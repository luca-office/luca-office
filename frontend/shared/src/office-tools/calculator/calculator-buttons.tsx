import {css} from "@emotion/react"
import * as React from "react"
import {Button} from "../../components"
import {IconName} from "../../enums"
import {fontColor} from "../../styles"
import {buttonHeight} from "../../styles/sizes"
import {CalculatorButtonConfig, calculatorButtonConfigs} from "./calculator-button-configs"
import {CalculatorButtonType} from "./enums/calculator-button-type"
import {CalculatorKey} from "./enums/calculator-key"
import {useCalculatorButtons} from "./hooks/use-calculator-buttons"
import {keyToIcon, keyToOperation} from "./utils/calculator-utils"

export interface CalculatorButtonsProps {
  readonly onButtonClick: (key: CalculatorKey) => void
  readonly isCalculatorFocussed: boolean
}

export const CalculatorButtons: React.FC<CalculatorButtonsProps> = ({onButtonClick, isCalculatorFocussed}) => {
  useCalculatorButtons({onButtonClick, isCalculatorFocussed})

  const renderButton = ({
    calculatorKey,
    type: buttonType = CalculatorButtonType.Primary,
    customStyles
  }: CalculatorButtonConfig) => {
    const iconName = keyToIcon(calculatorKey)
    const onClick = (evt: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
      // blur to prevent from being called on press of enter
      evt.currentTarget.blur()
      onButtonClick(calculatorKey)
    }
    return (
      <Button
        {...{
          key: calculatorKey,
          className: `operation-${calculatorKey.toLowerCase()}`,
          onClick,
          customStyles: [
            styles.button,
            buttonType === CalculatorButtonType.Secondary ? styles.buttonSecondary : null,
            calculatorKey === CalculatorKey.Divide ? styles.buttonSmallIcon : null,
            customStyles
          ],
          ...iconName
            .map<Partial<{icon: IconName; children: React.ReactNode}>>(icon => ({
              icon
            }))
            .getOrElse({
              children: keyToOperation(calculatorKey)
            })
        }}
      />
    )
  }

  return <div css={styles.grid}>{calculatorButtonConfigs.map(renderButton)}</div>
}

const styles = {
  grid: css({
    display: "grid",
    gridTemplateColumns: "repeat(5, 1fr)",
    gridTemplateRows: "repeat(5, 1fr)",
    gridGap: 8
  }),
  button: css({
    width: "100%",
    height: "100%",
    minHeight: buttonHeight,
    boxShadow: "0px 2px 2px 0px rgba(0, 0, 0, 0.11)",

    "> div > div": {
      marginRight: 0
    }
  }),
  buttonSecondary: css({
    background: "#f0f0f0",
    color: fontColor
  }),
  buttonSmallIcon: css({
    "svg > g > g": {
      strokeWidth: 0,
      fill: "white"
    }
  })
}
