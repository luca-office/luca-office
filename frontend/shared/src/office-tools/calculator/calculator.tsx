import {css} from "@emotion/react"
import * as React from "react"
import {OfficeWindow} from "../../components"
import {ViewerToolsType} from "../../enums"
import {CustomStyle, Flex, floatingWindowShadow, spacingMedium, zIndex3} from "../../styles"
import {CalculatorButtons} from "./calculator-buttons"
import {CalculatorDisplay} from "./calculator-display"
import {CalculatorKey} from "./enums"
import {useCalculatorHandling} from "./hooks/use-calculator-handling"

export interface CalculatorProps extends CustomStyle {
  readonly onKeyPress?: (key: CalculatorKey) => void
  readonly onClose: () => void
  readonly onMinimize?: () => void
  readonly onOutsideClick?: () => void
  readonly onMouseDown?: () => void
  readonly isFocussed: boolean
}

export const Calculator: React.FC<CalculatorProps> = ({
  customStyles,
  onClose,
  onMinimize,
  onKeyPress,
  onOutsideClick,
  onMouseDown,
  isFocussed
}) => {
  const {operation, result, handleOperation, hasInvalidInput} = useCalculatorHandling()

  const onOperationClick = (key: CalculatorKey) => {
    handleOperation(key)
    onKeyPress?.(key)
  }

  return (
    <OfficeWindow
      customStyles={[styles.window, customStyles]}
      toolType={ViewerToolsType.Calculator}
      isDraggable={true}
      draggableBoundsSelector={".desktop-content"}
      onClose={onClose}
      onMouseDown={onMouseDown}
      onMinimize={onMinimize}
      onOutsideClick={onOutsideClick}>
      <div css={styles.content}>
        <CalculatorDisplay hasInvalidInput={hasInvalidInput} operation={operation} result={result} />
        <div>
          <CalculatorButtons onButtonClick={onOperationClick} isCalculatorFocussed={isFocussed} />
        </div>
      </div>
    </OfficeWindow>
  )
}

export const CalculatorSizes = {
  window: {width: 344, height: 372}
}

const styles = {
  window: css({
    width: CalculatorSizes.window.width,
    height: CalculatorSizes.window.height,
    boxShadow: floatingWindowShadow,
    zIndex: zIndex3
  }),
  content: [
    Flex.column,
    css({
      padding: spacingMedium,
      background: "white"
    })
  ]
}
