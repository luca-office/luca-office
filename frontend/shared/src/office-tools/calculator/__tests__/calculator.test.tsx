import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {Button, OfficeWindow, ToolsHeader} from "../../../components"
import {Calculator, CalculatorProps} from "../calculator"

const defaultProps: CalculatorProps = {
  onKeyPress: jest.fn(),
  onClose: jest.fn(),
  isFocussed: true
}

const getComponent = (props?: Partial<CalculatorProps>) => <Calculator {...{...defaultProps, ...props}} />

describe("calculator", () => {
  it("renders correctly", () => {
    const component = getComponent()
    const tree = create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("calls onClose", () => {
    const mockOnClose = jest.fn()
    const component = getComponent({onClose: mockOnClose})
    const tree = shallow(component)

    const officeWindow = tree.find(OfficeWindow)
    expect(officeWindow).toHaveLength(1)

    const header = officeWindow.dive().find(ToolsHeader)
    expect(header).toHaveLength(1)

    const headerButtons = header.dive().find(Button)
    expect(headerButtons).toHaveLength(1)

    headerButtons.at(0).simulate("click")
    expect(mockOnClose).toHaveBeenCalledTimes(1)
  })
})
