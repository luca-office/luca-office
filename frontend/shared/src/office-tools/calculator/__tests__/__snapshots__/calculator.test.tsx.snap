// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculator renders correctly 1`] = `
<div
  className="react-draggable css-52xsf1-OfficeWindow"
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  style={
    {
      "transform": "translate(0px,0px)",
    }
  }
  transform={null}
>
  <div
    className="header css-hsza9j-OfficeWindow"
  >
    <div
      className="css-1sdua8l-ToolsHeader"
    >
      <div
        className="css-w2t5un-ToolsHeader"
      >
        <div
          className="css-1ss9q1q-Icon"
          dangerouslySetInnerHTML={
            {
              "__html": {},
            }
          }
        />
        <div
          className="css-1b4s2bm-Text"
        >
          calculator__label
        </div>
      </div>
      <div
        className="css-w2t5un-ToolsHeader"
      >
        <button
          className="css-1dzgoe2-Button"
          disabled={false}
          onClick={[Function]}
          title="viewer_tools__close_tool_label"
        >
          <div
            className="css-95j2hm-Button"
          >
            <div
              className="css-13zef26-Icon"
              dangerouslySetInnerHTML={
                {
                  "__html": {},
                }
              }
            />
          </div>
        </button>
      </div>
    </div>
  </div>
  <div
    className="css-1rjthu4-Calculator"
  >
    <div
      className="css-lb3e89-CalculatorDisplay"
    >
      <div
        className="css-evrz4g-Text"
      />
      <div
        className="css-0"
        id="tooltip_42"
      >
        <div
          className="css-1xmla1f-CalculatorDisplay"
          onClick={[Function]}
        >
          <div
            className="result css-1hjb42v-Text"
          >
            =
          </div>
          <div
            className="result css-lakr0j-Text"
          >
            0
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="css-19xuvvn-CalculatorButtons"
      >
        <button
          className="operation-openbracket css-1ejms2v-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            (
          </div>
        </button>
        <button
          className="operation-closebracket css-ozkjyz-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            )
          </div>
        </button>
        <button
          className="operation-ac css-1fytfbu-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            AC
          </div>
        </button>
        <button
          className="operation-seven css-1xvold1-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            7
          </div>
        </button>
        <button
          className="operation-eight css-1xvold1-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            8
          </div>
        </button>
        <button
          className="operation-nine css-1xvold1-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            9
          </div>
        </button>
        <button
          className="operation-divide css-wgy8l-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            <div
              className="css-kuxry5-Icon"
              dangerouslySetInnerHTML={
                {
                  "__html": {},
                }
              }
            />
          </div>
        </button>
        <button
          className="operation-reciprocal css-1fytfbu-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            1/x
          </div>
        </button>
        <button
          className="operation-four css-1xvold1-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            4
          </div>
        </button>
        <button
          className="operation-five css-1xvold1-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            5
          </div>
        </button>
        <button
          className="operation-six css-1xvold1-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            6
          </div>
        </button>
        <button
          className="operation-multiply css-1fytfbu-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            <div
              className="css-kuxry5-Icon"
              dangerouslySetInnerHTML={
                {
                  "__html": {},
                }
              }
            />
          </div>
        </button>
        <button
          className="operation-exponentiation css-1fytfbu-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            x^y
          </div>
        </button>
        <button
          className="operation-one css-1xvold1-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            1
          </div>
        </button>
        <button
          className="operation-two css-1xvold1-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            2
          </div>
        </button>
        <button
          className="operation-three css-1xvold1-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            3
          </div>
        </button>
        <button
          className="operation-subtraction css-1fytfbu-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            <div
              className="css-kuxry5-Icon"
              dangerouslySetInnerHTML={
                {
                  "__html": {},
                }
              }
            />
          </div>
        </button>
        <button
          className="operation-result css-1dfpps3-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            <div
              className="css-kuxry5-Icon"
              dangerouslySetInnerHTML={
                {
                  "__html": {},
                }
              }
            />
          </div>
        </button>
        <button
          className="operation-zero css-menk6j-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            0
          </div>
        </button>
        <button
          className="operation-separator css-1xvold1-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            .
          </div>
        </button>
        <button
          className="operation-addition css-1fytfbu-Button"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="css-95j2hm-Button"
          >
            <div
              className="css-kuxry5-Icon"
              dangerouslySetInnerHTML={
                {
                  "__html": {},
                }
              }
            />
          </div>
        </button>
      </div>
    </div>
  </div>
  <div
    className="css-pystkl-OfficeWindow"
  />
</div>
`;
